src/main/java/de/zalando/aruha/nakadi/FlowIdUtils.java

```diff
@@ -0,0 +1,44 @@
+package de.zalando.aruha.nakadi;
+
+import org.apache.log4j.NDC;
+
+import java.security.SecureRandom;
+import java.util.Base64;
+
+/**
+ * Helper class to deal with FlowIds.
+ *
+ * NOTE: It uses log4j's NDC and hence is bound to log4j (behind slf4j). If you change to another logging framework
+ * you have to rework this class.
+ */
+public final class FlowIdUtils {
+
+    public static final SecureRandom RANDOM = new SecureRandom();
+
+    private FlowIdUtils() {
+    }
+
+    public static void push(final String flowId) {
+        NDC.push(flowId);
+    }
+
+    public static String generateFlowId() {
+        final byte[] bytes = new byte[18];
+        RANDOM.nextBytes(bytes);
+        return Base64.getEncoder().encodeToString(bytes);
+    }
+
+    public static void clear() {
+        while (!"".equals(NDC.pop()));
+    }
+
+    public static String peek() {
```

Comment:
Yes.
If you create a new thread the FlowId stack from the parent thread is not passed to the child thread automatically. Hence we have to copy the stack to the child somehow. Or one hands over the parent thread's FlowId only.
There is a `copyStack` method missing. I will implement that as soon as Slava's branch is merged since it is a first and important case where we will need it (with the async streaming).